Sources/GRPC/GRPCKeepaliveHandlers.swift

```diff
@@ -0,0 +1,297 @@
+  /// Such a ping counts as a "ping strike".
+  private let minimumReceivedPingIntervalWithoutData: TimeAmount?
+
+  /// Maximum number of bad pings that the server will tolerate before sending an HTTP2 GOAWAY frame and closing the connection.
+  /// Setting it to `0` allows the server to accept any number of bad pings.
+  private let maximumPingStrikes: UInt?
+
+  /// When the last ping was received
+  private var lastReceivedPingDate: NIODeadline? = nil
+
+  /// When the last ping was sent
+  private var lastSentPingDate: NIODeadline? = nil
+
+  /// The number of pings sent on the transport without any data
+  private var sentPingsWithoutData = 0
+
+  /// Number of strikes
+  private var pingStrikes: UInt = 0
+
+  /// The scheduled task which will close the connection.
+  private var scheduledClose: Scheduled<Void>? = nil
+
+  /// Number of active streams
+  private var activeStreams = 0 {
+    didSet {
+      if activeStreams > 0 {
+        self.sentPingsWithoutData = 0
+      }
+    }
+  }
+
+  private let goAwayFrame = HTTP2Frame.FramePayload.goAway(lastStreamID: .rootStream, errorCode: .enhanceYourCalm, opaqueData: nil)
+
+  enum Action {
+    case schedulePing(delay: TimeAmount, timeout: TimeAmount)
+    case cancelScheduledTimeout
+    case reply(HTTP2Frame.FramePayload)
+  }
+
+  init(
+    pingCode: UInt64,
+    interval: TimeAmount,
+    timeout: TimeAmount,
+    permitWithoutCalls: Bool,
+    maximumPingsWithoutData: UInt,
+    minimumSentPingIntervalWithoutData: TimeAmount,
+    minimumReceivedPingIntervalWithoutData: TimeAmount? = nil,
+    maximumPingStrikes: UInt? = nil
+  ) {
+    self.pingCode = pingCode
+    self.interval = interval
+    self.timeout = timeout
+    self.permitWithoutCalls = permitWithoutCalls
+    self.maximumPingsWithoutData = maximumPingsWithoutData
+    self.minimumSentPingIntervalWithoutData = minimumSentPingIntervalWithoutData
+    self.minimumReceivedPingIntervalWithoutData = minimumReceivedPingIntervalWithoutData
+    self.maximumPingStrikes = maximumPingStrikes
+  }
+
+  func read(event: Any) -> Action? {
+    if event is NIOHTTP2StreamCreatedEvent && self.activeStreams == 0 {
+      self.activeStreams += 1
+      return .schedulePing(delay: self.interval, timeout: self.timeout)
+    } else if event is StreamClosedEvent {
+      self.activeStreams -= 1
+    }
+
+    return nil
+  }
+
+  func read(pingData: HTTP2PingData, ack: Bool) -> Action? {
+    let isPong = ack && pingData.integer == self.pingCode
+    let isIllegalWithoutCalls = !ack && self.activeStreams == 0 && !self.permitWithoutCalls
+    let isInvalidPing = !ack && isPingStrike
+    let isValidPing = !ack && !isPingStrike
+
+    if isPong {
+      return .cancelScheduledTimeout
+    } else if isIllegalWithoutCalls {
+      return .reply(self.goAwayFrame)
+    } else if isInvalidPing, let maximumPingStrikes = self.maximumPingStrikes {
+      if self.pingStrikes > maximumPingStrikes {
```

Comment:
If the maximum ping strikes is zero we should tolerate any number of strikes, that doesn't seem to be the case here?